drivers/dma-buf/dma-resv.c

```diff
@@ -498,6 +512,12 @@ int dma_resv_get_fences_rcu(struct dma_resv *obj,
 		ret = 0;
 unlock:
 		rcu_read_unlock();
+#ifdef __FreeBSD__
+		if (ret != 0) {
+			rw_rlock(&obj->rw);
```

Comment:
rw_lock here does not protect any data. writers are serialized with ww_mutex and readers are synchronized with writers with seqlock. What rw_lock does is blocking writers from spinning on seqlock if writer was preempted or blocked. If seqlock is used properly, writers can not block due to preempt_disable() section thus readers can just spin on seqlock. But in our case writer can be preempted, so it desirable to have blocking mechanism for readers.

Reply:
The pattern you use with rw_lock is very similar to that of SRCU in the LinuxKPI? Could you use SRCU instead of RW lock?

Reply:
The the writer section will be:
srcu_read_lock()
srcu_read_unlock()

And the rw_rlock/runlock will be synchronize_srcu().

You need to use "struct srcu_struct" and init that instead of the rw lock.

--HPS

Reply:
As I see, synchronize_srcu() waits for global SRCU grace period after srcu_read_unlock() has been issued, while rw_rlock/runlock unblocks immediately after rw_wunlock() has been called. I am not sure that we want such a delay here.

Reply:
I'm good with this approach. Has it been tested with a WITNESS enabled kernel?

Reply:
No witness warnings so far. That is expected as this change adds no new lock dependencies and LKPI forces disabling of witness with flags like MTX_NOWITNESS passed to locking functions.